projects/msal/fuzz_auth.py

```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/python3
+
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import atheris
+import sys
+with atheris.instrument_imports():
+    from msal import PublicClientApplication
+    from msal.application import extract_certs
+    from msal.authority import AuthorityBuilder
+
+def TestInput(input_bytes):
+    if len(input_bytes)<32:
+        return 
+
+    fdp = atheris.FuzzedDataProvider(input_bytes)
+
+    authority = AuthorityBuilder(fdp.ConsumeString(50),fdp.ConsumeString(50))
+
+    try:
+        app = PublicClientApplication(client_id=fdp.ConsumeString(32),authority=authority)
+        app.get_accounts()
+    except (ValueError,KeyError) as e:
+        error_list = [
+            "tenant_discovery_endpoint",
+            "Invalid IPv6 URL",
```

Comment:
Yes, some of them may come from low level python library which did happen when some unexpected input is passed from MSAL to them.

Reply:
the main point of catching these specific ones is that we want to catch other exceptions potentially thrown by MSAL, i.e. if MSAL throws a ValueError then we would like to see that.

In this context, is MSAL not expected to catch the errors thrown by the libraries it uses? In this sense I would suggest removing catching of the exceptions from lower down the stack and then asking MSAL to handle them -- unless it's specific in the documentation you have to handle it yourself

Reply:
> the main point of catching these specific ones is that we want to catch other exceptions potentially thrown by MSAL

That's understood. I have no problem of catching them here. I was just saying if an exception that you are trying to catch happens to have its specific type such as `ThatException`, then you can catch that exception by type, rather than by its message, regardless of where that exception was raised.

> In this context, is MSAL not expected to catch the errors thrown by the libraries it uses?
> suggest ... asking MSAL to handle them -- unless it's specific in the documentation you have to handle it yourself

I see your point. This is an open topic. We just haven't really decided on one way or another. The current behavior in MSAL is to not catch the 100 potential exceptions from underlying library Foo and rebrand them as MsalFooX, MsalFooY, MsalFooZ, and then wrap another 100 potential exceptions from underlying Bar...

And, nope, it is not specifically documented, but it seems fine in practice. I guess that is because (1) those lower level exception are usually rare, so rare that MSAL team does not encounter all of them; (2) when it happens, the raw exception being bubbled up is meaningful, so app developers are presumably OK with them.

Reply:
> I see your point. This is an open topic. We just haven't really decided on one way or another. The current behavior in MSAL is to not catch the 100 potential exceptions from underlying library Foo and rebrand them as MsalFooX, MsalFooY, MsalFooZ, and then wrap another 100 potential exceptions from underlying Bar...
> 
> And, nope, it is not specifically documented, but it seems fine in practice. I guess that is because (1) those lower level exception are usually rare, so rare that MSAL team does not encounter all of them; (2) when it happens, the raw exception being bubbled up is meaningful, so app developers are presumably OK with them.

Ah cool -- in that case then I think we should remove all of the checking of exceptions from lower-level libraries and let them be discovered as is now. That will also make it clear what can be thrown using MSAL